lib/logstash/inputs/snmp.rb

```diff
@@ -128,112 +129,115 @@ def register
-              result.update(data)
-            else
-              logger.debug? && logger.debug("walk operation returned no response", host: host, oid: oid)
-            end
-          rescue => e
-            logger.error("error invoking walk operation, ignoring", host: host, oid: oid, exception: e, backtrace: e.backtrace)
-          end
-        end
+      definition[:walk]&.each do |oid|
+        request.walk(target, OID.new(oid))
       end
 
-      if !Array(@tables).empty?
-        @tables.each do |table_entry|
-          begin
-            table_name = table_entry['name']
-            oids = table_entry["columns"].map { |oid| OID.new(oid) }
-            data = @client.table(target, table_name, oids)
-            if data&.any?
-              result.update(data)
-            else
-              logger.debug? && logger.debug("table operation returned no response", host: host, table: table_entry)
-            end
-          rescue => e
-            logger.error("error invoking table operation, ignoring",
-                         host: host, table_name: table_entry['name'], exception: e, backtrace: e.backtrace)
-          end
-        end
+      @tables&.each do |table_entry|
+        oids = table_entry['columns'].map { |oid| OID.new(oid) }
+        request.table(target, table_entry['name'], oids)
       end
 
-      unless result.empty?
-        event = targeted_event_factory.new_event(result)
-        event.set(@host_protocol_field, definition[:host_protocol])
-        event.set(@host_address_field, definition[:host_address])
-        event.set(@host_port_field, definition[:host_port])
-        event.set(@host_community_field, definition[:host_community])
-        decorate(event)
-        queue << event
-      else
-        logger.debug? && logger.debug("no snmp data retrieved", host: definition[:host_address])
+      max_host_timeout = definition[:timeout] if definition[:timeout] > max_host_timeout
+    end
+
+    in_flight_requests.each do |req|
+      definition = req[:definition]
+      result_consumer = lambda do |result|
+        if result&.any?
+          event = targeted_event_factory.new_event(result)
+          event.set(@host_protocol_field, definition[:host_protocol])
+          event.set(@host_address_field, definition[:host_address])
+          event.set(@host_port_field, definition[:host_port])
+          event.set(@host_community_field, definition[:host_community])
+          decorate(event)
+          queue << event
+        else
+          logger.debug? && logger.debug('No SNMP data retrieved', host: definition[:host_address])
+        end
       end
+
+      req[:request].get_result_async(result_consumer)
     end
+
+    # this timeout should be a safe-guard and it should never wait for that long.
+    # It might reach this point if there are several hosts operations, the hosts
+    # are unreachable/unhealthy, and/or the configured host timeout is too high.
+    in_flight_timeout = 1000 * 60 * 60 # 1 hour
```

Comment:
so theoretically, we could see this timeout be triggered when the input is polling 3 or 4 thousand devices. If each poll takes a second, then 3600 hosts would trip this.
Maybe we should have this upper limit configurable, or raise it much higher?

Reply:
Yes agreed! considering we don't actually know how fast/slow the operations might run, and if the `interval` is configured accordantly, with no overlapping.  I think having it configurable would be better. WDYT?

Reply:
++ on it being configurable.